File: code-external/matmul_dim.py
import random
import time

# Initialize the random number generator
random.seed(0)

# Generate two 800x800 matrices with random values
matrix_size = 500
matrix_a = [[random.uniform(0.0, 1.0) for _ in range(matrix_size)] for _ in range(matrix_size)]
matrix_b = [[random.uniform(0.0, 1.0) for _ in range(matrix_size)] for _ in range(matrix_size)]

# Measure the time it takes to perform the matrix multiplication
start = time.time()
result_matrix = [[0.0] * matrix_size for _ in range(matrix_size)]
for i in range(matrix_size):
    for j in range(matrix_size):
        for k in range(matrix_size):
            result_matrix[i][j] += matrix_a[i][k] * matrix_b[k][j]
duration = time.time() - start

# Print the time taken
print("Time taken: {:.6f} seconds".format(duration))

# Ensure the result is computed to avoid optimization removing the computation
print("Result (first element): {}".format(result_matrix[0][0]))

File: code-orig/matmul_dim.rs
extern crate ndarray;
use ndarray::prelude::*;
use rand::Rng;
use rand::distributions::Uniform; // Import Uniform from rand::distributions
use std::time::Instant;

fn main() {
    // Initialize the random number generator
    let mut rng = rand::thread_rng();

    // Generate two 800x800 matrices with random values
    let matrix_a: Array2<f64> = Array2::from_shape_fn((500, 500), |_| rng.sample(Uniform::new(0.0, 1.0)));
    let matrix_b: Array2<f64> = Array2::from_shape_fn((500, 500), |_| rng.sample(Uniform::new(0.0, 1.0)));

    // Measure the time it takes to perform the matrix multiplication
    let start = Instant::now();
    let result_matrix: Array2<f64> = matrix_a.dot(&matrix_b);
    let duration = start.elapsed();

    // Print the time taken
    println!("Time taken: {:?}", duration);

    // Ensure the result is computed to avoid optimization removing the computation
    println!("Result (first element): {}", result_matrix[(0, 0)]);
}
